### ISSUE TITLE
Ability to Search Word Vectors

### ISSUE THREAD

**Issue by mdavis95** [lumongo/lumongo#116]
Store a document with normal fields and the vector of the document:

{
  title: "Cancer and cancer awareness",
  abstract: "Cancer awareness is important...",
  vec: [0.03, 0.021, 0.04, ..]
}

Search for similar documents using a documents vector or a vector computed by a piece of text.  The average word vector of a document or a piece of text can be used or more complex methods.

Ideally the comparison would involve the cosine similarity of the search vector versus all the documents in the corpus but this is too slow.  Instead use SuperBit (https://papers.nips.cc/paper/4847-super-bit-locality-sensitive-hashing.pdf) as an estimate of cosine similarity.  Java implementation of SuperBit is here (https://github.com/tdebatty/java-LSH).  Index the SuperBit of the word vector one field per bit.  When querying use a minimum should match query on the SuperBit of the query vector.   The amount of bits that need to match given a cosine similarity can be determined with the following formula:

```java
int mm = (int) ((1 - (Math.acos(cosineSimRequest.getSimilarity()) / Math.PI)) * signature.length);
```

**Comment by mdavis95** [comment]
050cd6b07c248bafa684b4e4505c8a4fda309174
d0149f8946180798ef7b34685c74f3526e2b7ad8
663cb3f10dcae7a6b01141ba0710662a32b010c6
ba4b691d7ffec246a3a2b4208e4aa48458ba328c

**Comment by mdavis95** [comment]
Simple example java code

At create/update index time
```java
indexConfig.addFieldConfig(FieldConfigBuilder.create("vec", FieldType.NUMERIC_DOUBLE).projectAsSuperBit("vec", 300));
```

At indexing time
```java
List<Double> vec = new ArrayList<>();
document.put("vec", vec);
```

At query time
```java
double[] vec = [];
double sim = 0.9;
query.addCosineSim("vec", vec, sim);
```

**Comment by mdavis95** [comment]
a5df4a964961daa2df54dfe05f6913de1ef5d3cd

**Comment by mikemccand** [comment]
Could you also use Lucene's `MoreLikeThis` class?  It's designed for this use case, where you pass a prototype document and it finds documents with similar terms.

**Comment by mdavis95** [comment]
I have implemented similar functionality to more like this in a two step process. 1) Grab the documents and get the top N TFIDF weighted terms 2) Do a search with the terms and a minimum should match

Howver, at my office we do a lot of work with word vectors so I was trying to find a way to make them searchable.  The super bit approach works well but on a corpus of 25 million publications the query takes around 40 seconds.  I encoded a word vector of 300 doubles in my testing into 900 bits.  It seems that a minimum should match query on 900 bits where mm~880 is not the fastest in lucene.
